Lib/test/test_clinic.py

```diff
@@ -1692,5 +1692,167 @@ def test_permute_optional_groups(self):
+            ("\na\nb\n\n",      "a\nb"),
+            ("\n\na\nb\n\n",    "a\nb"),
+            ("\n\na\n\nb\n\n",  "a\n\nb"),
+            # Note, leading whitespace is preserved:
+            (" a\nb",               " a\nb"),
+            (" a\nb ",              " a\nb"),
+            (" \n \n a\nb \n \n ",  " a\nb"),
+        )
+        for lines, expected in dataset:
+            with self.subTest(lines=lines, expected=expected):
+                out = clinic.strip_leading_and_trailing_blank_lines(lines)
+                self.assertEqual(out, expected)
+
+    def test_normalize_snippet(self):
+        snippet = """
+            one
+            two
+            three
+        """
+        template = (
+            "{indent}one\n"
+            "{indent}two\n"
+            "{indent}three"
+        )
+        for indent in 0, 4, 8:
+            expected = template.format(indent=" " * indent)
+            with self.subTest(snippet=snippet, indent=indent, expected=expected):
+                out = clinic.normalize_snippet(snippet, indent=indent)
+                self.assertEqual(out, expected)
+
+    def test_accumulator(self):
+        acc = clinic.text_accumulator()
+        self.assertEqual(acc.output(), "")
+        acc.append("a")
+        self.assertEqual(acc.output(), "a")
+        self.assertEqual(acc.output(), "")
+        acc.append("b")
+        self.assertEqual(acc.output(), "b")
+        self.assertEqual(acc.output(), "")
+        acc.append("c")
+        acc.append("d")
+        self.assertEqual(acc.output(), "cd")
+        self.assertEqual(acc.output(), "")
+
+    def test_c_repr(self):
+        dataset = "", "a", "abc"
+        for line in dataset:
+            expected = f'"{line}"'
+            with self.subTest(line=line, expected=expected):
+                out = clinic.c_repr(line)
+                self.assertEqual(out, expected)
+
+    def test_quoted_for_c_string(self):
+        dataset = (
+            # input,    expected
+            (r"abc",    r"abc"),
+            (r"\abc",   r"\\abc"),
+            (r"\a\bc",  r"\\a\\bc"),
+            (r"\a\\bc", r"\\a\\\\bc"),
+            (r'"abc"',  r'\"abc\"'),
+            (r"'a'",    r"\'a\'"),
+        )
+        for line, expected in dataset:
+            with self.subTest(line=line, expected=expected):
+                out = clinic.quoted_for_c_string(line)
+                self.assertEqual(out, expected)
+
+    def test_rstrip_lines(self):
+        lines = (
+            "a \n"
+            "b\n"
+            " c\n"
+            " d \n"
+        )
+        expected = (
+            "a\n"
+            "b\n"
+            " c\n"
+            " d\n"
+        )
+        out = clinic.rstrip_lines(lines)
+        self.assertEqual(out, expected)
+
+    def test_format_escape(self):
+        line = "{}, {a}"
+        expected = "{{}}, {{a}}"
+        out = clinic.format_escape(line)
+        self.assertEqual(out, expected)
+
+    def test_indent_all_lines(self):
```

Comment:
Similar thoughts to my comments above on `test_normalize_snippet`. If you're computing the expected result in your test, then maybe you need to write a test for your test to make sure the test is computing the expected result correctly ;)

I'd go for something a little more verbose but more declarative, that involves less computation in the test, and where it's easier for humans to see at-a-glance what the expected result is.